### PR TITLE
Adding the new CTE correction to the pipeline

### PR DESCRIPTION
**Pull request by julienguy**
In the PR, the CTE correction is included in the pipeline.

* New script ```desi_fit_cte_night``` and new I/O for CTE parameters
  - The script is to be run per night after the afternoon calibs in daily ops
  - It looks for LED 1s and 120s exposures in the exposure tables
  - preprocess them with a different output file (```preproc-cte-CAM-NIGHT.fits```)
  - Runs Eddie's model fit on the data for cameras with keywords ```CTECOLSX``` with X being an amplifier id (A,B,C,D)
  - writes output tables ```$DESI_SPECTRO_CALIB/$SPECPROD/calibnight/NIGHT/ctecorr-CAMERA-NIGHT.csv```
   (or for each red and NIR camera by default)

* Changes to desi_proc
  - ```desi_proc -n NIGHT --nightlycte``` runs ```desi_fit_cte_night -n NIGHT``` in the pipeline
  - The batch version is not implemented (```desi_proc -n NIGHT --nightlycte --batch``` crashes)

* Changes to desi_preproc
  - Cameras with keywords ```CTECOLSX``` have a new CTE correction applied at the preprocessing level
  - Uses row-by-row Horn extraction to get a precise image model
  - Uses the nightly CTE parameters

Requires modifications to the ```DESI_SPECTRO_CALIB``` yaml files. One can use the following version and try the code

```
export DESI_SPECTRO_CALIB=/global/cfs/cdirs/desi/users/jguy/teststand/desi_spectro_calib
export SPECPROD=XXXX
desi_fit_cte_night -n 20231111
desi_preproc -n 20231111 -e 204371 --camera z3
```

### CODE REVIEW

File: py/desispec/correct_cte.py [hunk 1]
```diff
@@ -8,12 +8,14 @@
...
-from desispec.io import findfile, read_image
-from desispec.io.xytraceset import read_xytraceset
-from desispec.io.fiberflat import read_fiberflat
+import desispec.io
+import desispec.io.util
+import desispec.io.xytraceset
+import desispec.io.fiberflat
 from desispec.trace_shifts import compute_dx_from_cross_dispersion_profiles
-from desispec.preproc import get_amp_ids, parse_sec_keyword
+import desispec.preproc
```
Comment (schlafly) [inline py/desispec/correct_cte.py:18]: Glad this helped!

File: py/desispec/scripts/proc.py [hunk 2]
```diff
@@ -273,7 +282,44 @@ def main(args=None, comm=None):
...
+            timer.start('nightlycte')
+
+            camword = create_camword(args.cameras)
+            cmd = f"desi_fit_cte_night -n {args.night} -c {camword}"
+
+            log.info(f'RUNNING {cmd}')
+
+            #- Note: nightly_bias may not produce all biasnight files if some
+            #- are determined to be worse than the default, so check existence
+            #- of output files separately.
```
Comment (schlafly) [inline py/desispec/scripts/proc.py:302]: Is this comment vestigial?  nightly_bias seems irrelevant here, though I admit to limited understanding of what proc.py is trying to do.

**Review by schlafly — APPROVED** [review]
Thanks, looks good.  I left a few minor comments inline.  This is a useful demo of what one needs to do to put a new kind of calib "in the pipeline."  Perhaps a minimal sketch would be, just for my reference:
- make a CLI for running the calib
- add an entry to io.meta for the new calib file type
- update proc.py to know how to call the CLI (MPI support would go here)
- set up triggering so that calibs get made each night (not in this PR)

File: py/desispec/correct_cte.py [hunk 3]
```diff
@@ -437,11 +490,24 @@ def fit_cte(images):
         par = least_squares(chi, [startguesses[bestguess], 0.2],
                             diff_step=[0.2, 0.01], loss='huber')
         ctefits[ampname] = dict()
-        offcols = f'{tcte["start"]}:{tcte["stop"]}'
+        offcols = f'{tcte["start"]}:{tcte["stop"]}:Y'
```
[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/correct_cte.py:493]: good catch. I did that change when running but forgot to commit this one

File: py/desispec/scripts/proc.py [hunk 4]
```diff
@@ -273,7 +282,44 @@ def main(args=None, comm=None):
...
+            result, success = runcmd(desispec.scripts.fit_cte_night.main,
+                    args=cmd.split()[1:], inputs=[], outputs=[], comm=comm)
+
+            #- check for ctecorrnight
+            ctecorrnighfiles = [findfile('ctecorrnight', args.night, camera=cam) for cam in args.cameras]
+            for filename in ctecorrnighfiles:
```
[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/scripts/proc.py:308]: done

**Review by sbailey — CHANGES_REQUESTED** [review]
I've posted some comments inline, many of which I have flagged for myself to implement, in particular adding MPI support for the pipeline.

One non-inline comment since I can't seem to comment on things that pre-dated this PR: `desispec.correct_cte.correct_image` code appears to require a `cteparams` input, and it is in the docstring, but it isn't explicitly in the function call signature so I don't see how it even runs and it appears that the function isn't used anymore.  Has it been fully replaced by `correct_image_via_model`?

**Comment by schlafly** [comment]
> One non-inline comment since I can't seem to comment on things that pre-dated this PR: `desispec.correct_cte.correct_image` code appears to require a `cteparams` input, and it is in the docstring, but it isn't explicitly in the function call signature so I don't see how it even runs and it appears that the function isn't used anymore. Has it been fully replaced by `correct_image_via_model`?

Thanks.  This was my fault; the function should have had a cteparams input.  As you suggest, the problem is that we don't actually use that mode, and instead use correct_image_via_model, and the two interfaces have fallen out of sync.  We could just delete correct_image, though I think that approach is actually more correct in the limit that rowbyrow == spectroperfectionism.  I did not go back to exploring that approach after implementing the model & rowbyrow components, though.

File: py/desispec/correct_cte.py [hunk 5]
```diff
@@ -177,50 +175,81 @@ def get_amps_and_cte(image, cteparam=None):
...
+    night = desispec.preproc.header2night(header)
+    camera = header['CAMERA'].lower()
+
+    if with_params :
+        # look for CTE param table for this camera
+        filename = desispec.io.findfile('ctecorrnight', night=night, camera=camera)
+        log.debug(f"Looking for file {filename}")
+        if not os.path.isfile(filename) :
+            log.debug(f"No CTE file {filename}")
+            return dict(),dict()
```
[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/correct_cte.py:190]: My idea was that the code would support both options: saving or not an empty CSV table. But I will rewrite the code such that if there is a CTECOLS parameter in the calib but no CSV table then the code would throw an error. (and also if this table exists but does not contain the correct entry). I will commit after testing at NERSC when computer is back online.

[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/correct_cte.py]: change tested and committed

File: py/desispec/correct_cte.py [hunk 6]
```diff
@@ -319,6 +348,13 @@ def chi_simplified_regnault(param, cleantraces=None, ctetraces=None,
     return res.reshape(-1)
 
 
+def get_transfer_function(function_name) :
```
[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/correct_cte.py:371]: committed

File: py/desispec/sky.py [hunk 7]
```diff
@@ -121,25 +121,26 @@ def get_sector_masks(frame):
...
-                    mess="cannot decode {}={}".format(key,val)
-                    log.error(mess)
-                    raise KeyError(mess)
-                xb = max(sec[1].start,int(tmp2[0]))
-                xe = min(sec[1].stop,int(tmp2[1]))
-                sector = [yb,ye,xb,xe]
-                sectors.append(sector)
-                log.info("Adding CCD sector in amp {} with offset: {}".format(
-                    amp,sector))
+        for key in [ "OFFCOLS"+amp , "CTECOLS"+amp ] :
```
[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/sky.py:133]: committed

File: py/desispec/correct_cte.py [hunk 8]
```diff
@@ -462,20 +528,47 @@ def get_cte_images(night, camera):
...
-    # before that one??
-    onesecexp = ((np.abs(exptable['EXPTIME'] - 1) < 0.1) &
-                 (exptable['OBSTYPE'] == 'flat'))
-    index = np.flatnonzero(onesecexp)[0]
-    imagefn = [
-        findfile('preproc', night, exptable['EXPID'][i], camera)
-        for i in (index, index - 1)]
-    images = [read_image(fn) for fn in imagefn]
+
+    index1 = np.where(((np.abs(exptable['EXPTIME'] - 1) < 0.1) & (exptable['OBSTYPE'] == 'flat')))[0][0]
```
[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/correct_cte.py]: committed

File: py/desispec/correct_cte.py [hunk 9]
```diff
@@ -462,20 +528,47 @@ def get_cte_images(night, camera):
...
+
+    # get header and primary header of image
+    filename  = desispec.io.findfile('raw',night,exptable['EXPID'][index1])
+    header    = fitsio.read_header(filename, camera.upper())
+    amp, cte = get_amps_and_cte(header,with_params=False)
+    if len(cte)==0 :
+        log.info(f"No CTE correction to compute for {night} {camera}")
+        return None
+
+    index2 = np.where(((np.abs(exptable['EXPTIME'] - 120) < 10) & (exptable['OBSTYPE'] == 'flat')))[0][-1]
```
[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/correct_cte.py]: committed

File: py/desispec/preproc.py [hunk 10]
```diff
@@ -1300,6 +1304,16 @@ def preproc(rawimage, header, primary_header, bias=True, dark=True, pixflat=True
 
     img = Image(image, ivar=ivar, mask=mask, meta=header, readnoise=readnoise, camera=camera)
 
+    #- CTE correction if any exist
+    if not no_cte_corr :
+        if not ( 'DESI_SPECTRO_REDUX' in os.environ and 'SPECPROD' in os.environ ) :
+            log.warning("No DESI_SPECTRO_REDUX or no SPECPROD defined. Cannot find calibration data, so cannot do a CTE correction")
```
[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/preproc.py:1310]: We have the same issue already with the nightly bias. The code won't crash if it's missing the CTE params. I'm adding the extra option to read a calib file ...

[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/preproc.py]: committed

**Comment by sbailey** [comment]
As a heads up, I'm strongly considering changing this to output a single file per night instead of 20 files per night, where currently only 2 of them have actual corrections (z1,z3) and in much of the past none of them would have corrections.  I think the simplification from fewer files (e.g. when combining files across nights to make trends) outweighs the under-the-hood complexity of gathering and stacking tables to write a single file.  I will investigate that as part of adding MPI support for the pipeline (maybe that under-the-hood complexity is worse than I thought).

File: py/desispec/correct_cte.py [hunk 11]
```diff
@@ -151,16 +153,12 @@ def add_cte(img, cte_transfer_func=None, **cteparam):
...
+    """Get amp and CTE information from image header.
 
     Parameters
     ----------
-    image : object
-        Must have 'meta' attribute.
-    cteparam : dict or None
-        dictionary of form {'z1C': {'543:2057': (115.0, 0.21)}} that
-        provides the CTE parameters for the trap on a particular image.
+    header (can be image.meta)
```
[the remark this answers is not in the document]
Comment (sbailey) [reply py/desispec/correct_cte.py]: For the record, I split this function into two: `get_amp_regions_to_cte_correct` to get the regions that need to be corrected (used by `fit_cte`) and `get_cte_params` to actually get the parameters output by an earlier run of `fit_cte`.

`get_amp_regions_to_cte_correct` is equivalent to previous `get_amps_and_cte(..., with_params=False)` and `get_cte_params` is equivalent to previous `get_amps_and_cte(..., with_params=True)`.  Maybe that was just cosmetic, but splitting the functionality into different functions helped me understand what they were doing.

**Comment by sbailey** [comment]
Summary of my changes:
  * added MPI support
  * output one ctecorr file per night instead of per camera per night
  * some semi-cosmetic rearrangement and docstring updates to help me understand the structure for long term maintainability
  * minor bug fixes like creating output directory if it didn't already exist

The remaining piece is actually integrating with desi_proc_daily_manager and desi_run_night, which I'll need to discuss with @akremin .  We currently start submitting jobs after the 300s DARK at the end of the ZEROs, before the ARCs and FLATs.  Now we need to wait for the 1s FLAT before starting to process anything, but the current structure doesn't have hooks for "notice that we have received an ARC/FLAT that we need to process but wait until we get more data before submitting it because it will depend upon processing future data that may or may not eventually exist..."

Let's wait to merge this until we have a plan for that final integration, because merging this requires a 3-way coordination between desispec, $DESI_SPECTRO_CALIB, and nightwatch.

**Comment by sbailey** [comment]
Update:
  * I've added desi_fit_cte_corr to the ccdcalib job, including a check to not run it if the ctecorr file already exists e.g. due to linking to a previous night.
  * I've added a "wait_for_cals" step to the beginning of desi_daily_proc_manager to wait for the end of calibration files to arrive before starting any processing, which enables it to use the flats in the ccdcalib job (which was previously launched as soon as zeros+darks were done, before arcs+flats had arrived).

This is getting increasingly fragile, but enables us to run with the new CTE calibrations while waiting for a better pipeline cleanup from @akremin .  I'll run in a separate side prod for a few nights to ensure that it really works before merging into main.

**Comment by sbailey** [comment]
My test run last night patiently waited until all of the calibrations had arrived before trying to process any of them, and then immediately fell on its face trying to CTE correct a dark image for identifying bad columns before the nightly ctecorr existed.  I fixed that by updating desi_proc to *not* try to CTE correct ZEROs and DARKs.

@schlafly @julienguy if we get into a situation where the CTE corrections are large enough that we need to worry about their impact on dark current, we'll need to revisit this.

Before merging, I'm going to keep running the test prod mimicking daily ops until we have a night with actual on-sky science data to confirm that also gets CTE corrected correctly.

**Comment by schlafly** [comment]
That sounds right to me.

I hadn't thought about CTE in conjunction with darks and zeros before.  I guess it's actually pretty important for hot pixels in both cases; since all the traps are empty those must get smeared badly.  But we really don't have any mechanism to do the right thing there.  The CTE modeling code uses a "model image" that would be ~all zero there, or something.  The previous code that did not use a model image at the expense of correlating the read noise would conceptually do the right thing if somehow we already had the CTE parameters.  But correlating the read noise is bad and would be most of what you end up doing on the darks and zeros.

I guess there's a world where when generating the master darks we have enough data that the read noise could be averaged down, and we have the past CTE parameters and could do something.  But that's another project; I agree we shouldn't go there now!

**Comment by akremin** [comment]
In the ccdcalib job instead of running the cte detection after the bad
column detection, we could run it before so that it is available for the
dark.

If up-to-date bad columns are crucial we could have the cte detection
create temporary versions of those files like it does the preproc of the
flats.

The second portion of the suggestion would make a complicated procedure
even more complicated, but I thought I'd throw it out there.

**Comment by sbailey** [comment]
Yeah, all options are messy albeit in different ways.  I'm going to leave it as-is for now unless we have a demonstrated case that it is worth the effort to change.

**Comment by sbailey** [comment]
@schlafly @SGontcho I'd like to put this into production on Thursday night, or otherwise wait until next week so that we're not making major changes over the weekend.  This also requires a matching Nightwatch update from @sybenzvi to call preproc with --no-cte-corr.  I updated that for qproc here (it will never try to ctecorr, because that isn't q=quick), but there is also a preproc-only mode for nightwatch that needs to be updated (same place in redrock.run.run_preproc that desihub/nightwatch#380 updated).

**Review by sbailey — APPROVED** [review]

**Comment by sbailey** [comment]
There was a bug in preproc where --model-variance was replacing the cte-corrected image back to the original one (thanks to @julienguy for identifying that).  I've fixed that, and verified that:
  * for a non-CTE camera, this PR matches main with and without --model-variance
  * for a CTE-corrected camera (z3), this PR
    * matches main with --no-cte-correction
    * meaningfully differs from main in the CTE-corrected region regardless of --model-variance or not
  * --model-variance only changes ivar, but not the pixel image.

I'm re-launching night 20240128 in the /global/cfs/cdirs/desi/users/sjbailey/spectro/redux/ctetest production for testing...

**Comment by sbailey** [comment]
@schlafly @SGontcho @julienguy I've reprocessed /global/cfs/cdirs/desi/users/sjbailey/spectro/redux/ctetest night 20240128 (but not the other nights).

* The z3 preproc file now differ only in the CTE-corrected region (upper left).
* This leaks into the extracted raw frames at other wavelengths and fibers (lower left), which is perhaps a side-effect of differences in trace shifting the PSF before extraction?  The scale is larger than the "divide-and-conquer" extraction sub-regions.
* This further propagates into the calibrated frames (cframe) differences (lower right), where the impact of both sky fibers and standard stars understandably impacts fibers other than just the CTE corrected ones.
* Redshifts are shown in upper right.  There are ~15 differences, all coming from ZWARN>0 targets (red).
* I've verified that there are no differences on non-CTE corrected cameras, but I'd appreciate additional eyes on these results, especially the impact on non-CTE corrected fibers.

image

Plotting code is at /global/cfs/cdirs/desi/users/sjbailey/spectro/redux/ctetest/compare_exp.py

**Comment by schlafly** [comment]
Here's the tileqa comparison I usually make:
https://data.desi.lbl.gov/desi/users/schlafly/misc/schlafly-20240128-ctetest-tileqa.pdf
On this night I really only see an important effect on the redshifts in tileqa on 22161, the one you show, and it's pretty modest.  But it is a definite improvement.
image
image
Note the patch of blue target redshifts around z ~ 1 in daily that disappear in ctetest.

I hadn't thought about impacts on PSF traces before, but I agree that we expect changes at more wavelengths after the sky subtraction, etc., starts correlating everything.

**Comment by sbailey** [comment]
Thanks for the QA check, @schlafly.

FWIW, I confirmed that the frame differences (before any sky subtraction or flux calibration that would correlate wavelengths and fibers) are almost entirely due to the difference in the PSF trace shifts.  i.e. extracting the ctetest preproc image using the daily trace-shifted PSF results in frame differences mostly in the CTE-corrected region as expected:

image

There is a smattering of other isolated differences that I'm still not sure about, but it might be edge effects on the divide-and-conquer from how it was called in the pipeline vs. my isolated test.

I think we're good to go for putting this into production on Monday.

**Comment by schlafly** [comment]
I would also support putting it into production---thanks!

**Comment by sbailey** [comment]
I tagged DESI_SPECTRO_CALIB 0.5.1 as the last version *before* this change, then committed the updated spec/sm10/sm10-z.yaml (z1) and spec/sm6/sm6-z.yaml files (z3).  Both DESI_SPECTRO_CALIB trunk and desispec/main have been updated at NERSC.

https://github.com/desihub/nightwatch/pull/390 is the matching PR from Nightwatch.

**Pull request merged**